python/cudf/cudf/core/reshape.py

```diff
@@ -496,16 +508,33 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
                 # Explicitly cast rather than relying on None being falsy.
                 sort=bool(sort),
             )
+            if keys is not None:
+                result.index = cudf.MultiIndex.from_tuples(
+                    [
+                        (k, i)
+                        for k, x in zip(keys, [obj.shape[0] for obj in objs])
+                        for i in range(x)
+                    ]
+                )
```

Comment:
will look into this, thanks

Reply:
@wence- if the following  is not correct, can you please provide an example of what would be correct?
```python
In [40]: pd.concat(x, axis=0)
Out[40]: 
            A    B    C
first  a  1.0  3.0  NaN
       b  2.0  4.0  NaN
second 1  5.0  7.0  NaN
       2  6.0  8.0  NaN
third  d  NaN  NaN  1.0
       g  NaN  NaN  2.0
       h  NaN  NaN  3.0
```

Reply:
The pandas code looks right, but I think the iteration over `range(x)` in the patch ignores the existing index and would instead just produce some concatenated range indices?

For the example shown, I think we have:

```
keys = ["first", "second", "third"]
```

and
```
objs = [pd.DataFrame(**{"data": {"A": [1, 2], "B": [3, 4]}}, index=["a", "b"]),
        pd.DataFrame(**{"data": {"A": [5, 6], "B": [7, 8]}}, index=["c", "e"]),
        pd.DataFrame(**{"data": {"C": [1, 2]}}, index=["d", "g"])]
```

So
```
                    [
                        (k, i)
                        for k, x in zip(keys, [obj.shape[0] for obj in objs])
                        for i in range(x)
                    ]
```

Produces:
```
[('first', 0),
 ('first', 1),
 ('second', 0),
 ('second', 1),
 ('third', 0),
 ('third', 1),
 ('third', 2)]
```

So the first level is correct, but the inner level is wrong (it should be `"a", "b", "c", "e", "d", "g"`).

Instead, if we build the level indices level-by-level as suggested above, we can produce the correct inner level.

Does that make sense?

Reply:
confirming -- by inner level you mean index `[1]` within each `tuple`, so instead of:
```python
[('first', 0),
 ('first', 1),
 ('second', 0),
 ('second', 1),
 ('third', 0),
 ('third', 1),
 ('third', 2)]
```

we want:
```python
[('first', "a"),
 ('first', "b"),
 ('second', "c"),
 ('second', "e"),
 ('third', ??),
 ('third', "d"), ??
 ('third', "g")] ??
```

Reply:
this is what i'm seeing from pandas, which makes sense to me:
```python
>>> x = {
...     "first": (pd.DataFrame(**{"data": {"A": [1, 2], "B": [3, 4]}, "index":["a", "b"]})),
...     "second": (pd.DataFrame(**{"data": {"A": [5, 6], "B": [7, 8]}, "index":["c", "e"]})),
...     "third": (pd.DataFrame(**{"data": {"C": [1, 2]}, "index":["d", "g"]})),
... }
>>> 
>>> z= pd.concat(x, axis=0)
>>> z
            A    B    C
first  a  1.0  3.0  NaN
       b  2.0  4.0  NaN
second c  5.0  7.0  NaN
       e  6.0  8.0  NaN
third  d  NaN  NaN  1.0
       g  NaN  NaN  2.0
>>> z.index
MultiIndex([( 'first', 'a'),
            ( 'first', 'b'),
            ('second', 'c'),
            ('second', 'e'),
            ( 'third', 'd'),
            ( 'third', 'g')],
           )
```

Reply:
@wence- i've updated the code to produce the same `MultiIndex` index as `pd`. i think it's simpler than the route you were suggesting.

Reply:
the `axis=0` and `tuple` columns are still an issue, so i still need to sort that out

Reply:
Thanks, I see. If we can, I'd like to keep the construction of the new multiindex levels on device rather than coming back to the host and iterating over things.

Consider the case where each dataframe has a few million rows, it would be nice not to construct these tuples individually and then go back to the GPU.

That's what the more complicated code I sketched is trying to do:

`libcudf.filling.repeat` is like `numpy.repeat` but for cudf columns rather than numpy arrays. And `concat_columns` is the low-level concatenation of cudf columns (which is what we need for the "inner" levels)